src/main/java/net/minecraftforge/common/lighting/LightingEngine.java

```diff
@@ -0,0 +1,630 @@
+    private EnumSkyBlock lightType;
+
+    //Iteration state data
+    //Cache position to avoid allocation of new object each time
+    private final MutableBlockPos curPos = new MutableBlockPos();
+    private PooledLongQueue curQueue;
+    private Chunk curChunk;
+    private long curData;
+
+    //Cached data about neighboring blocks (of tempPos)
+    private boolean isNeighborDataValid = false;
+    private final Chunk[] neighborsChunk = new Chunk[EnumFacing.VALUES.length];
+    private final MutableBlockPos[] neighborsPos = new MutableBlockPos[EnumFacing.VALUES.length];
+    private final int[] neighborsLight = new int[EnumFacing.VALUES.length];
+
+    public LightingEngine(final World world)
+    {
+        this.world = world;
+        this.profiler = world.theProfiler;
+
+        for (int i = 0; i < EnumSkyBlock.values().length; ++i)
+        {
+            this.queuedLightUpdates[i] = new PooledLongQueue();
+        }
+
+        for (int i = 0; i < this.queuedDarkenings.length; ++i)
+        {
+            this.queuedDarkenings[i] = new PooledLongQueue();
+        }
+
+        for (int i = 0; i < this.queuedBrightenings.length; ++i)
+        {
+            this.queuedBrightenings[i] = new PooledLongQueue();
+        }
+
+        for (int i = 0; i < this.neighborsPos.length; ++i)
+        {
+            this.neighborsPos[i] = new MutableBlockPos();
+        }
+    }
+
+    /**
+     * Schedules a light update for the specified light type and position to be processed later by {@link #procLightUpdates(EnumSkyBlock)}
+     */
+    public void scheduleLightUpdate(final EnumSkyBlock lightType, final BlockPos pos)
+    {
+        this.scheduleLightUpdate(lightType, posToLong(pos));
+    }
+
+    /**
+     * Schedules a light update for the specified light type and position to be processed later by {@link #procLightUpdates()}
+     */
+    private void scheduleLightUpdate(final EnumSkyBlock lightType, final long pos)
+    {
+        final PooledLongQueue queue = this.queuedLightUpdates[lightType.ordinal()];
+
+        queue.add(pos);
+
+        //make sure there are not too many queued light updates
+        if (queue.size() >= MAX_SCHEDULED_COUNT)
+        {
+            this.procLightUpdates(lightType);
+        }
+    }
+
+    /**
+     * Calls {@link #procLightUpdates(EnumSkyBlock)} for both light types
+     */
+    public void procLightUpdates()
+    {
+        this.procLightUpdates(EnumSkyBlock.SKY);
+        this.procLightUpdates(EnumSkyBlock.BLOCK);
+    }
+
+    /**
+     * Processes light updates of the given light type
+     */
+    public void procLightUpdates(final EnumSkyBlock lightType)
+    {
+        final PooledLongQueue queue = this.queuedLightUpdates[lightType.ordinal()];
+
+        if (queue.isEmpty())
+        {
+            return;
+        }
+
+        //renderer accesses world unsynchronized, don't modify anything in that case
+        if (this.world.isRemote && !Minecraft.getMinecraft().isCallingFromMinecraftThread())
```

Comment:
If the METHOD is resolved. Then all classes ret linked. So the current method would get a class not found exception if it is either called, or it's reflected over.

Take the following for example:
`class Test {
  public void someMethod() {
    if (!remote && ClientClass.func()) {}
  }
}`
If someone CALLS that, or Does Test.class.getDeclaredMethods() then it will throw a ClassNotFoundException

Now take this:
`class Test {
  public void someMethod() {
    if (!remote && clientBouncer()) {}
  }
  public boolean clientBouncer(){
    return ClientClass.func();
  }
}`
Now, if someone calls `someMethod` then it will NOT throw an error, as long as remote is true.
That's because the JVM is lazy-linked {In all JVM implementations i've seen}. So when it compiles the 'someMethod' function is only cares that it's calling clientBouncer and it lazy compiles so it will only compile clientBouncer if the logic ever causes the function to be called.
Now, Reflection throws a wrench in the works. Getting the Method/MethodHandle for a method causes the JVM to compile that method, so doing Test.class.getDeclaredMethods would throw an error because it's trying to compile clientBouncer and running into a client only method.

Get it?

As for ForgeHooks, that class is not reflected over, and the only place onPickBlock is called is from specifically the Minecraft class which is purely client code, so we can get away with it. If some other mod comes along and reflects all the methods in ForgeHooks they will crash, and THEY will be responsible for it. 

As for what the JLS states, there is a MAJOR difference between what it states and what actually exists in the real world. I deal in reality not specs. However, if you REALLY want to be safe, like I said, you could use the FML*Handler classes, this type of case is exactly what they are designed for.